Fix: no-implicit-coercion string concat false positive (fixes #7057)

kaicataldo opened the pull request:
The rule was checking that the `Literal` is not a number in the case of `+foo` (so it would not warn for `+42`), but was not doing the same for strings.

TheSavior commented:
Thanks for doing this!

platinumazure commented on lib/rules/no-implicit-coercion.js:
```diff
@@ -106,14 +106,34 @@ function getNonNumericOperand(node) {
+ * string literal or not.
+ */
+function isString(node) {
+    return node.type === "Literal" && typeof node.value === "string";
```
Consider using [`isStringLiteral` from ast-utils](https://github.com/eslint/eslint/blob/master/lib/ast-utils.js#L229)? (One significant difference, that function also checks for TemplateLiterals. I think that might be okay here, though.)

kaicataldo replied:
Oh cool! Didn't know about this - thanks!

platinumazure commented on tests/lib/rules/no-implicit-coercion.js:
```diff
@@ -64,6 +64,10 @@ ruleTester.run("no-implicit-coercion", rule, {
 
+        // https://github.com/eslint/eslint/issues/7057
+        {code: "'' + 'foo'"},
+        {code: "+42"},
```
Could you add a valid test for `"str += "non-empty"`? I didn't see it in the test file but you did change `isAppendEmptyString`. Thanks!

platinumazure commented:
Left a suggestion about using an ast-utils method and adding a test.

In addition: If you do follow my suggestion and use ast-utils' `isStringLiteral` method, since that also allows TemplateLiteral nodes, could you add some representative tests that use template strings in place of strings in existing valid/invalid tests?

Thanks, otherwise looks great!

kaicataldo commented:
Updated with suggested changes - thanks!

platinumazure commented:
Thanks @kaicataldo!

I was originally thinking of tests like this:

```js
foo + ``;
`` + foo;
```

Maybe I'm missing something-- are empty TemplateLiterals syntactically invalid?

kaicataldo commented:
This is something that the rule does not currently check - do we want to lump that change into this bug fix?

kaicataldo commented:
Updated - this has become a sort of enhancement in addition to the bug fix, in that it is now updated to work with ES6 `TemplateLiterals`. I'm fine with doing the bug fix and enhancement in the same PR, but would like to make sure others are okay with it too since it wasn't discussed in the issue.

platinumazure commented:
Sorry, I didn't realize templates weren't flagged by this!

On Sep 5, 2016 6:03 PM, "ESLint Bot" <notifications@github.com> wrote:

> LGTM
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/eslint/eslint/pull/7060#issuecomment-244819693>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AARWej94oj8U_lvyaw6-PX7rYGlbi-pKks5qnJ_EgaJpZM4J1PEI>
> .
>

kaicataldo commented:
Seems like a worthwhile update (I think this rule was written before ES6), just want to make sure it doesn't accidentally get snuck in without consideration :)

Happy to split this up into two PR's if we decide we want to keep it separate.

TheSavior commented:
It seems like the easier and less controversial approach would be to split this into two PRs.

platinumazure commented:
@kaicataldo Sorry, this came up because I wanted you to use ast-utils. :smile: If you want to revert that change so this PR remains a bugfix, please go ahead. I'll write an enhancement request issue for enforcing on template literals as well.

kaicataldo commented:
No need to apologize :)

Alright, let's split this up. Will revert to the original PR and we can discuss adding support for template strings in a separate issue.

kaicataldo commented:
Reverted to just the original bug fix (but still using `ast-utils.isStringLiteral()` - I still would prefer to use the built-in utility function with an extra guard than re-implement). I checked out a new branch with the `TemplateLiteral` check and tests, so if we decide to implement that it'll be quite simple!

kaicataldo commented:
@platinumazure How does this look now? Want to make sure I get your feedback since you commented on the original iteration!

platinumazure commented:
LGTM, thanks!

vitorbal commented:
LGTM

ilyavolodin commented:
LGTM. Thanks
